### PR TITLE
Add a setting max_execution_time_leaf to limit the execution time on shard for distributed query

### LINKED ISSUE #46102: Add max_execution_time_leaf
**Use case**
Currently there is max_execution_time setting, that can be set in subqueries. But it makes it hard to use this setting with distributed sub-queries, as it is likely that timeout will be reached before distributed subquery finished. 

**Describe the solution you'd like**

Add max_execution_time_leaf settings similar to the other restrictions on query complexity that only apply to distributed sub queries.

**Describe alternatives you've considered**

It is possible to workaround this using features like view, or dyanmically with cluster(view()) to set manually timeout in the subquery since we have control over them and they are not generated internally. But it is much less handy than simply using distributed tables, and for some reason other max_* seettings have a leaf version.

### PR DESCRIPTION
**Pull request by canhld94**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add a setting `max_execution_time_leaf` to limit the execution time on shard for distributed query, and `timeout_overflow_mode_leaf` to control the behaviour if timeout happens.

Close #46102 

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by canhld94** [comment]
@alexey-milovidov @ucasfl can we consider adding this settings? I think it's also reasonable to have this settings when we only want to limit the execution time on remote shard, not on coordinator.

**Comment by alexey-milovidov** [comment]
Sorry, we forgot about this pull request - we will pick it up.

File: src/Interpreters/ClusterProxy/SelectStreamFactory.cpp [hunk 1]
```diff
@@ -120,17 +122,41 @@ void SelectStreamFactory::createForShard(
...
+        query = query_ast->clone();
+        /// Replace 'max_execution_time' of this sub-query with 'max_execution_time_leaf' and 'timeout_overflow_mode'
+        /// with 'timeout_overflow_mode_leaf'
+        if (auto * select = query->as<ASTSelectQuery>())
+        {
+            auto query_settings = select->settings();
+            if (!query_settings)
+                query_settings = std::make_shared<ASTSetQuery>();
+            
+            query_settings->as<ASTSetQuery &>().is_standalone = false;
```
Comment (alexey-milovidov) [inline src/Interpreters/ClusterProxy/SelectStreamFactory.cpp:140]: Interesting... Usually, we send the modified settings in the server protocol, not inline in the query. I will check if this method of sending settings is viable.

**Comment by alexey-milovidov** [comment]
@canhld94 I've resolved conflicts and fixed typos, but more changes are needed.

File: src/Interpreters/ClusterProxy/SelectStreamFactory.cpp (hunk 1 above)
Comment (Avogar) [reply src/Interpreters/ClusterProxy/SelectStreamFactory.cpp:140]: Why not to change these settings in [updateSettingsForCluster](https://github.com/ClickHouse/ClickHouse/blob/aa7c1f63abda4cdb8a592c9d90dc3a7d2ad3a935/src/Interpreters/ClusterProxy/executeQuery.cpp#L38)? Is there a reason why we cannot do it there?

Comment (canhld94) [reply src/Interpreters/ClusterProxy/SelectStreamFactory.cpp]: @Avogar I changed the implementation as your suggestion.

**Review by Avogar — APPROVED** [review]

File: src/Core/Settings.h [hunk 2]
```diff
@@ -384,8 +384,10 @@ class IColumn;
...
+    M(Seconds, max_execution_time, 0, "If query runtime exceeds the specified number of seconds, the behavior will be determined by the 'timeout_overflow_mode', which by default is - throw an exception. Note that the timeout is checked and query can stop only in designated places during data processing. It currently cannot stop during merging of aggregation states or during query analysis, and the actual run time will be higher than the value of this setting.", 0) \
     M(OverflowMode, timeout_overflow_mode, OverflowMode::THROW, "What to do when the limit is exceeded.", 0) \
+    M(Seconds, max_execution_time_leaf, 0, "Similar semantic to max_execution_time but only apply on leaf node for distributed queries, the time out behavior will be determined by 'timeout_overflow_mode_leaf' which by default is - throw an exception", 0) \
+    M(OverflowMode, timeout_overflow_mode_leaf, OverflowMode::THROW, "What to do when the leaf limit is exceeded.", 0) \
```
Comment (Avogar) [inline src/Core/Settings.h:390]: Let's also add these settings to documentation near `max_execution_time` [here](https://github.com/ClickHouse/ClickHouse/blob/master/docs/en/operations/settings/query-complexity.md#max_execution_time-max-execution-time)

**Pull request merged**